### PR TITLE
Refactor DB maintenance functionality into package

### LINKED ISSUE #693: Move db-related functions out of Quicksetup
The [Quicksetup](https://github.com/aiidateam/aiida_core/blob/59e5f39fbc24b456ea29a5c54539855dffce5d48/aiida/cmdline/verdilib.py#L591) class contains several database-related functions (e.g. for setting up profiles) that are useful for other commands besides `verdi quicksetup` - perhaps even beyond the cli.

They should be moved to a more general level, such as the base-level of [aiida/cmdline/verdilib.py](https://github.com/aiidateam/aiida_core/blob/59e5f39fbc24b456ea29a5c54539855dffce5d48/aiida/cmdline/verdilib.py) or perhaps even [aiida/common/setup.py](https://github.com/aiidateam/aiida_core/blob/59e5f39fbc24b456ea29a5c54539855dffce5d48/aiida/common/setup.py) (I'm not an expert of the code structure).

Furthermore, there are no tests for these functions as of yet.

### PR DESCRIPTION
**Pull request by DropD**
## Still to be done: 
- [x] add documentation for the new `control.postgres` module.
- [x] add pre-commit extras requirements: `pre-commit`, `yapf`, `prospector`, `pylint`
- [x] add documentation for `pre-commit` usage

## What changed
The functions that allow the `verdi quicksetup` and the `verdi profile delete` commands to create / drop database users and databases have been refactored into their own module and given a public API which is accessed through the new `aiida.control.postgres.Postgres` class.

### Why not in common?
I believe that common should be for code that is used internally by different parts of `AiiDA`, not by specialized, public APIs that happen to be used internally too in very few places

### Why a new submodule?
* This is the first effort to make some of the logic behind the command line into a user facing API to better support scripting / building tools around AiiDA. A lot more will follow and to avoid cluttering up other modules and to help users / script writers find these APIs quickly, I feel all of that should be in a separate module
* A separate module allows to put in place code style / quality checks for the new code without interrupting other code or introducing a lot of formatting changes.

## Motivation
* Will fix #693
* Cleaner design
* Provide a public, easy to find API for writing tools in / around AiiDA

## Introducing pre-commit (only for new code)
Introduces [pre-commit](http://pre-commit.com/): a hook that checks for code sanity and applies the google code style (official style of the project), ONLY in the newly created `control` module. This is to keep code quality and consistency high.

### How it works
pre-commit allows to run tests on the code, both as a pre commit git hook, as well as in CI
* APPLIES ONLY TO THE NEWLY CREATED `aiida.control` MODULE
* every push will be tested in a separate CI stage. If the tools find coding mistakes or the style does not conform to the official style, the builld will fail.
* after downloading `pre-commit` and the necessary linter / formatter tools, one can install the git hook (`cd aiida-core && pre-commit install`). This will run the code quality tests on every commit (only on changed files, so it is quick).
* the formatter hook will actually take care of the formatting, all you have to do if it finds anything is to `git add` the changes it makes and committing again.
* it prevents you from most easy to make (copy pasta etc) mistakes before you embarrass yourself by publishing them.

### CODE REVIEW

**Comment by ltalirz** [comment]
Sorry for the long wait, I was extremely busy.
This looks fantastic, both the separation of the postgres functions and the linter!

I just tested the linter and it promptly told me I made a "pointless-string-statement" :)
I will become an avid user of this functionality.

I guess, for security reasons, it isn't possible to have it install the git pre-commit hooks automatically when one selects the `dev_precommit` extra(?). Perhaps one could print the remaining commands `cd aiida-core && pre-commit install`  after the installation on the command line?
We should think about every barrier we can remove for people to start using it, including how to enable it for the part of code they are working on.

I will go through the code later today.

**Comment by DropD** [comment]
Another thing to keep in mind is that it is not necessary to install the pre-commit hook if you are not going to touch the `aiida.control` module. As of yet. So it may be a little early to tell everyone to install it right away.

However, i agree that after installing the `dev_precommit` extra one should be notified to run `pre-commit install`.

**Comment by ltalirz** [comment]
Yes, obviously this only applies, when you select the `dev_precommit` extra.

I don't know what others think @giovannipizzi  but once we have worked with the linter for a bit (say, we've cleaned up one existing module) and are happy with the settings, in my opinion it should apply to all new code (or at least new files) by default.

File: .coveragerc [hunk 1]
```diff
@@ -0,0 +1,7 @@
+[run]
+data_file = coverage/.coverage
```
Comment (ltalirz) [inline .coveragerc]: I cannot find this data_file, is it something everybody has to provide by themselves?

File: .travis.yml [hunk 2]
```diff
@@ -50,28 +42,42 @@ install:
...
+              # Here I setup the test AiiDA profile, non-interactively
+              - verdi -p test_$TEST_AIIDA_BACKEND setup --non-interactive --backend=$TEST_AIIDA_BACKEND --email="aiida@localhost" --db_host="localhost" --db_port=5432 --db_name="test_$TEST_AIIDA_BACKEND" --db_user=postgres --db_pass='' --repo="/tmp/test_repository_test_${TEST_AIIDA_BACKEND}/" --first-name=AiiDA --last-name=test --institution="AiiDA Team" --no-password
+
+              # Maybe not needed, but set this profile to be the default one for (at least) the daemon
+              - verdi profile setdefault daemon $TEST_AIIDA_BACKEND
+              - verdi profile setdefault verdi $TEST_AIIDA_BACKEND
+
+              - .travis-data/before_script.sh
+
+          script: .travis-data/test_script.sh
```
Comment (ltalirz) [inline .travis.yml]: This is tested, right? I haven't used travis so far, I just know that we shouldn't break it ;-)

File: .coveragerc (hunk 1 above)
Comment (DropD) [reply .coveragerc]: This is somewhat separate from the rest of the PR: I added sensible defaults for the `coverage` tool. The `data_file` is the one that will be created when you run for example

```bash
coverage run /path/to/verdi -p test_profile devel tests && coverage html -i
```

I set it to be in a folder with all the additional report / annotation files that `coverage` can generate.

Long term I will try to add a coverage report stage to travis. VERY long term we might have it fail if it gets too low ;)

File: .travis.yml (hunk 2 above)
Comment (DropD) [reply .travis.yml]: The same tests are run as before ([see here](https://travis-ci.org/DropD/aiida_core/builds/275702922) or click on 'Show checks' and then 'Details' below). The only difference is that now two stages exist, one for pre-commit, one for tests. If pre-commit fails tests are not run to save time.

**Comment by ltalirz** [comment]
I just have a final question: As a developer adding new stuff, how do I decide whether it goes into `common` or `control` (or a different module)?
Can we find two reasonably short sentences, saying:
`common` contains ...
`control` contains ...

These sentences could then go as docstrings into the respective __init__.py

Related to this is also the question of the names - personally if I look at the source code and see `control` and `common`, it doesn't tell me a lot about what these modules contain (most of the other module names are significantly more specific).
But that's a minor issue and I'm fine with keeping the names if we don't come up with something clearly better.

**Comment by DropD** [comment]
Actually, the docs should contain a description of what the control module contains, too!

* `common` contains internal and somewhat general functionality that is reused in many parts of AiiDA
* `control` contains user facing APIs to control various aspects of AiiDA from scripts / tools / verdi cli

I know that this means some code from `common` will end up in `control` eventually, sometimes with some work on the interfaces.

**Comment by ltalirz** [comment]
Great, what about this iteration:

* `common` contains internal functionality that is needed by multiple modules of AiiDA
* `control` contains user facing APIs to control AiiDA from the verdi cli, scripts or plugins

As an AiiDA user, from this I would understand that I can expect the APIs in `control` to be more stable and that I should first try using them before diving into `common`.

**Comment by DropD** [comment]
I have no objections. I am unsure about where to include that information. Other than the `control` and `common` docs. Or is that enough?

File: aiida/control/postgres.py [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+
+        postgres = Postgres()
+        postgres.determine_setup()
+        postgres.create_dbuser('username', 'password')
+        if not postgres.db_exists('dbname'):
+            postgres.create_db('username', 'dbname')
+
+    Complex Example::
+
+        postgres = Postgres(port=4321, interactive=True)
```
[the remark this answers is not in the document]
Comment (DropD) [reply aiida/control/postgres.py]: Done

**Comment by ltalirz** [comment]
I personally would always look first inside the `__init__.py` of the corresponding modules.
It also means that this information will be automatically present in the API documentation generated by sphinx.

**Pull request merged**